### DISCUSSION TITLE
💖 extending number of ansible-lint cores (maintainers)

### DISCUSSION THREAD

**Discussion by ssbarnea** [ansible/ansible-lint#1325] [General] [answered]
##### Summary
I would like to propose extending the list of people with core rights (review + merge) on ansible-lint project.

##### Additional Information

Mainly the **community maintenance** was an expected measure when the project was adopted about 1.5 years ago. Since then, @webknjaz lead the project and lots of changes where made. Still, maintenance can easily become a burden and no project should rely on a single person, especially and that is mostly his personal time. It is much better to have a small group of people doing this.

Regarding how to achieve this goal, we could use a a layered approach. First we invite those interested to help to become **collaborators** with triage permissions. Likely those would need at least one or two contributions to the project in the last year before being considered collaborators.

This means that they would be able to edit labels, assign tickets, close/open tickets and PRs and perform reviews. Still, they would not be able to push/merge directly.

Once they prove to be reliable reviewers, we can elevate their permissions to allow merges.

Regarding merges we can also enable the "require 1 review for merge" option, so nobody would be able to merge stuff without having at least someone else approve his change. For me this is a no-brainer even for small projects, as on OpenStack we require at least two reviews, and reviewers need to be from different companies. Luckily we don't need such requirements here.

The 3rd level of access is the permission to create a release (tag), that is something where we should continue to rely on Ansible Galaxy decision, especially for major version changes.

**Comment by gundalow** [comment, marked as answer]
@ssbarnea Thank you for raising this.
I'm in favour of giving extra people permissions if they want them. GitHub's permissions model has been expanded over the years. Also I assume people will act in good faith, so I'm happy for people to be given higher permissions, then can always been removed in the future if need be.

@ssbarnea Are you happy for people to use this issue to track who maybe interested in being added to the project? Do you have any initial suggestions you'd be happy sharing?

**Comment by ssbarnea** [comment]
I would like to nominate https://github.com/albinvass as I worked with him on maintaining [zuul-roles](https://review.opendev.org/#/q/project:zuul/zuul-jobs) (ansible), he made few good contributions, including writing a custom rule for zuul project.

If others want to involve more in shaping the future of ansible-lint too, now it would be a good moment to address their interest.

**Comment by albinvass** [comment]
> I would like to nominate https://github.com/albinvass as I worked with him on maintaining zuul-roles (ansible), he made few good contributions, including writing a custom rule for zuul project.

I would be willing to take a maintainer role. But I already have a ton to do so I'm not entirely sure how active I would be able to be. :)

**Comment by ssbarnea** [comment]
Apparently the only thing guaranteed around ansible-lint is delaying making any decisions regarding its maintenance and proliferating the culture of cowboy coding.

While I do have a lot of respect about the time @webknjaz invested into this project, I cannot say the same thing about the practice of making changes without any reviews and fully ignoring CI/CD results. 

Take a look below, no reviewers, lots of jobs failing. That is not a unique case, in fact is how most of the changes where made.

![](https://sbarnea.com/ss/Screen-Shot-2020-06-12-09-04-19.98.png)

Yep, mistakes can be made by anyone but when we do know is extremely easy to prevent such mistakes by enforcing checks on github project settings. Even if these where asked multiple times over the last two years, nothing happened.

> The road to hell is paved with good intentions

I am still waiting for the the github invitation to become collaborator (not core), but I kinda lost hope.

For those that may be surprised why this escalated like this is because the ansible-lint maintenance issue was an open subject since it was adopted by Ansible team, discussions happened on irc, tickets.

**Comment by gundalow** [comment]
@ssbarnea and @albinvass now both have commit on ansible-lint
All PRs now require a +1 from another maintainer.

**Comment by ssbarnea** [comment]
@gundalow Thanks for that! This move will help making the linter an actively maintained and evolving. Both of us have good experience working in distributed projects with very strict review and merge polices.

I want to state that the invitation will remain open for others. We do want to see others interested in becoming core contributors to the project. In fact we do not need much help implementing new features, most useful help is to review open reviews and occasional proposals that could affect the behavior of the tool.

**Comment by dpendolino** [comment]
I'd be happy to help out, I use ansible-lint personally and professionally and would love give back.

**Comment by webknjaz** [comment]
@dpendolino feel free to start with contributing to https://github.com/ansible/ansible-lint/labels/help%20wanted. Not having privileges on GitHub shouldn't stop you from contributing...
Meanwhile, let's ask @gundalow to add you with the "Triage" access level if that's possible.

**Comment by gundalow** [comment]
@dpendolino Welcome to the community, thank you for offering to help. 

For long winded reason, github.com/ansible only has permissions for read/write/owner. So for the moment I've given you `read` permissions. This will allow others to assign issues and PRs to you, please do shout out if there are any you'd like assigning to yourself.

GitHub is used to track work. We also use `#ansible-lint` on Freenode IRC to chat with each other.

Thanks again, and please do shout out if there is anything we can do to help support you.

**Comment by webknjaz** [comment]
@dpendolino FYI you'll need to accept the invite first (check your email inbox!) so that we'll be able to assign you to issues and reviews.

**Comment by tohtohro** [comment]
Hi, I would like to help out too! As mentioned previously, I have working knowledge with python and am trying to build up on my ansible knowledge at the same time. Will like to contribute to this project since i have the time too.

**Comment by dpendolino** [comment]
@webknjaz Accepted!

**Comment by webknjaz** [comment]
@tohtohro great! Do you need any pointers? You could start with things under "help wanted" that include some refactoring requests, for example. Also, participating in reviews is a good place for help too.

**Comment by nixfu** [comment]
I am a huge ansible-lint fan and think it is one of the key tools that ansible users have.  I would be happy to help out with triage or whatever else is needed.  I am a fairly experienced python programmer as well.

**Comment by webknjaz** [comment]
@nixfu perfect! There's some tasks related to the project maintainability (like getting rid of unittest.TestCase relics) and just bugs (that may need reproducers or confirming against master):
- https://github.com/ansible/ansible-lint/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%22help+wanted%22
- https://github.com/ansible/ansible-lint/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Abug

Choose whatever you care about and participate :)

**Comment by benyanke** [comment]
I'd be happy to jump in as a maintainer of ansible-lint. It's a very useful tool, and has helped me improve my playbooks.

**Comment by webknjaz** [comment]
@benyanke feel free to start contributing then :) I'll ask @gundalow to invite you too...

**Comment by gundalow** [comment]
@nixfu and @benyanke I'm glad you find ansible-lint useful, thank you for your kind offer of help. If you accept your invite via https://github.com/ansible/ansible-lint/invitations then we can assign issues and PRs to you. Just let us know which you'd like to work on. In the future, we should be able to give you `triage` access, though at the moment that isn't available for github/ansible/*

**Comment by gundalow** [comment]
@tohtohro Same as above for you. Thank you :)

**Comment by webknjaz** [comment]
@nixfu @tohtohro don't forget to accept the invitation or we won't be able to assign you to tasks or ask for reviews in PRs.

**Comment by rafaelfolco** [comment]
I'd be more than happy to help with reviews. Please let me know how do I get involved. Thanks!

**Comment by greg-hellings** [comment]
*raises hand* Put me in, coach! I spend my days writing Ansible and Molecule tests - and half of my evenings, as well.

**Comment by gundalow** [comment]
@rafaelfolco @greg-hellings Welcome and thank you for your kind offer to help. Please accept the invite via https://github.com/ansible/ansible-lint/invitations then we can assign issues and PRs to you. Just let us know which you'd like to work on. In the future, we should be able to give you `triage` access, though at the moment that isn't available for github/ansible/* yet.

**Comment by ssbarnea** [comment]
I would like to propose @geerlingguy and @greg-hellings as cores as I seen both as being being focused on Ansible QA and will a good eye to spot potential issues that could be caused by incoming changes. 

It also happens that both of them are redhatters, which should ease the vetting process.

**Comment by geerlingguy** [comment]
I'd be okay with it, though I can't promise any level of commitment. I can jump in in bursts here and there as I get time :) (as I'm sure we'd all say, ha!).

**Comment by ansiblejunky** [comment]
@ssbarnea I am interested in helping out here. I did some analysis on lint tools last year and covered some of the "features" and have some ideas on direction. Let me know how I can help out.

**Comment by ssbarnea** [comment]
Watch the project, perform reviews and feel free to try making any PR. I would focus on simple changes for which we already have tickets.

Also be sure you join the irc channels, they are a very good way to collaborate.

Look at existing PRs that are not drafts, regardless if CI reports red, sometimes that red comes from a non-voting check. Look at them and comment, feel free to agree or disagree with anything you see there.

In case it was not clear, merging any change requires one peer core review. This works fine for changes coming from outside as we managed to close all but it can prove more problematic if changes are proposed by am existing core, as there are less options available, especially if consensus is harder to reach. Still, even if you are not a core, you can review and explain why you support or not a particular change. Usually this should be enough in order to unblock PRs.

**Comment by ssbarnea** [comment]
@gundalow @awcrosby I would to propose transferring ansible-lint project from ansible github organization to ansible-community organization, like we did back in January with molecule.

This will allow us to invite other collaborators to the project and make its maintenance easier.

The only downside I see is that the github url will be bit longer, but I think that the benefits will be seen very quickly. 

Also the transfer does not break any URLs as github is very good at redirecting.

**Comment by ssbarnea** [comment]
Any update on this? We should also consider dropping those that did not had any activity in 6 months from the list of cores, so the list of potential reviewers would be decluttered. Obviously that their status could easily be restored if they return but it does make no sense to list them otherwise.

**Comment by apatard** [comment]
Sorry for hi-jacking a little bit but from my ansible-lint user point of view, as long as it's maintained, finding ansible-lint in ansible-community instead of ansible doesn't make a lot of difference. So, if the only way to see it maintained is to move it, so be it. I mean, for instance, molecule is used even if it's in ansible-community/ Again, all that really matters is to have it maintained and not dead.

**Comment by xoxys** [comment]
@ssbarnea Thanks for inviting me.

**Comment by ssbarnea** [comment]
I am closing this to cleanup the backlog but keep in mind that the invitation to extend the list of maintainers will always be open.

**Comment by webknjaz** [comment]
@ssbarnea how about converting this into a discussion? I think it belongs there + it could be pinned on that page.

**Reply by ssbarnea** [reply]
Done! I guess that is a discussion that will never be closed as new contributors and maintainers are more than welcomed.

**Comment by ssbarnea** [comment]
@ssato I did send you an invite to our team as I seen your recent pull request and previous interest in improving ansible linting. For the moment is only with triage access but after few more pull requests I will also enable write access.

**Reply by ssato** [reply]
@ssbarnea Thanks a lot for your invitation!

**Comment by ssbarnea** [comment]
I would like to propose making @cognifloyd as core on ansible-lint as a way to show my appreciation for [his current contributions](https://github.com/ansible-community/ansible-lint/commits?author=cognifloyd) and very interesting pending ones too. I looked at his code and I am confident that he will not only with new features but also improving the code quality. Feel free to upvote/downvote my proposal.

**Reply by cognifloyd** [reply]
Cool. Thank you! Happy to help.

**Comment by IronTooch** [comment]
I'd be happy to do some of the basic stuff, like tagging, opening/closing tickets, and commenting, to help reduce the noise for the folks doing the heavy lifting.

**Reply by ssbarnea** [reply]
You are more than welcome to start at any time. I would recommend you to start small, like with a very small contribution fix, maybe improving the docs or rule description? That is only to get you warm-up, as I seen you want to add a stats-report feature which is great (i wanted to do it myself for very long time).

I will send you an invite after you make the first contribution.

**Comment by ssbarnea** [comment]
Please note that we will perform some housekeeping activities and remove write access for those that were not active in a long period of time, like more than an year.

That is a security measure and be assured that we will be more than happy to see you back.
